File: trace/trace_test.go
package trace

import (
	"context"
	"net/http"
	"strings"
	"testing"
	"time"
)

// Shared fixtures, hoisted so the sample ids and source aren't repeated literals
// across the tables (keeps goconst quiet and the intent obvious).
const (
	sampleTraceID   = "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31"
	sampleRequestID = "018f3a1c-2b4d-7e8f-9a0b-1c2d3e4f5061"
	sampleSource    = "runpod-graphql"
)

func TestValidID(t *testing.T) {
	tests := []struct {
		description string
		input       string
		expected    bool
	}{
		{description: "canonical uuid v4", input: sampleTraceID, expected: true},
		{description: "uuid v7 style", input: sampleRequestID, expected: true},
		{description: "service-prefixed id with underscore", input: "trace_018f3a1c2b4d7e8f", expected: true},
		{description: "dotted id", input: "svc.abc.123", expected: true},
		{description: "all allowed classes", input: "AZaz09-_.", expected: true},
		{description: "exactly maxIDLen chars", input: strings.Repeat("a", maxIDLen), expected: true},
		{description: "empty string is rejected", input: "", expected: false},
		{description: "one over maxIDLen is rejected", input: strings.Repeat("a", maxIDLen+1), expected: false},
		{description: "embedded CRLF (header/log injection) is rejected", input: "abc\r\nX-Evil: 1", expected: false},
		{description: "bare newline is rejected", input: "abc\ndef", expected: false},
		{description: "space is rejected", input: "abc def", expected: false},
		{description: "control byte is rejected", input: "abc\x00def", expected: false},
		{description: "non-ascii is rejected", input: "abcé", expected: false},
		{description: "slash is rejected", input: "abc/def", expected: false},
		{description: "colon is rejected", input: "abc:def", expected: false},
	}
	for _, tt := range tests {
		t.Run(tt.description, func(t *testing.T) {
			if got := validID(tt.input); got != tt.expected {
				t.Errorf("validID(%q) = %v, want %v", tt.input, got, tt.expected)
			}
		})
	}
}

func TestResolveID(t *testing.T) {
	tests := []struct {
		description   string
		input         string
		wantPropagate bool // true: returned verbatim; false: freshly generated
	}{
		{description: "valid id is propagated verbatim", input: sampleTraceID, wantPropagate: true},
		{description: "absent id is generated", input: "", wantPropagate: false},
		{description: "oversized id is regenerated", input: strings.Repeat("a", maxIDLen+1), wantPropagate: false},
		{description: "CRLF id is regenerated", input: "abc\r\nInjected: 1", wantPropagate: false},
		{description: "junk id is regenerated", input: "abc def", wantPropagate: false},
	}
	for _, tt := range tests {
		t.Run(tt.description, func(t *testing.T) {
			got := resolveID(tt.input)
			if tt.wantPropagate {
				if got != tt.input {
					t.Errorf("resolveID(%q) = %q, want it propagated verbatim", tt.input, got)
				}
				return
			}
			if got == tt.input {
				t.Errorf("resolveID(%q) returned the input; expected a fresh id", tt.input)
			}
			if !validID(got) {
				t.Errorf("resolveID(%q) generated an invalid id %q", tt.input, got)
			}
		})
	}
}

func TestResolveSource(t *testing.T) {
	tests := []struct {
		description string
		input       string
		expected    string
	}{
		{description: "empty source stays empty", input: "", expected: ""},
		{description: "valid service name is kept", input: sampleSource, expected: sampleSource},
		{description: "CRLF source is dropped to empty", input: "svc\r\nX-Evil: 1", expected: ""},
		{description: "spaced source is dropped to empty", input: "not a slug", expected: ""},
	}
	for _, tt := range tests {
		t.Run(tt.description, func(t *testing.T) {
			if got := resolveSource(tt.input); got != tt.expected {
				t.Errorf("resolveSource(%q) = %q, want %q", tt.input, got, tt.expected)
			}
		})
	}
}

func TestFromHeaderOrNew(t *testing.T) {
	const goodTrace = sampleTraceID
	const goodReq = sampleRequestID

	tests := []struct {
		description       string
		headers           map[string]string
		wantTraceID       string // "" => expect a generated valid id
		wantRequestID     string // "" => expect a generated valid id
		wantTraceSource   string
		wantRequestSource string
	}{
		{
			description:       "all ids present and valid are propagated",
			headers:           map[string]string{headerTraceID: goodTrace, headerRequestID: goodReq, headerTraceSource: "main-ui", headerRequestSource: "hapi"},
			wantTraceID:       goodTrace,
			wantRequestID:     goodReq,
			wantTraceSource:   "main-ui",
			wantRequestSource: "hapi",
		},
		{
			description:   "no headers generates both ids",
			headers:       map[string]string{},
			wantTraceID:   "",
			wantRequestID: "",
		},
		{
			description:   "poisoned trace id is regenerated, valid request id kept",
			headers:       map[string]string{headerTraceID: "abc\r\nX-Evil: 1", headerRequestID: goodReq},
			wantTraceID:   "",
			wantRequestID: goodReq,
		},
		{
			description:     "poisoned source is dropped to empty",
			headers:         map[string]string{headerTraceID: goodTrace, headerRequestID: goodReq, headerTraceSource: "svc\r\nX-Evil: 1"},
			wantTraceID:     goodTrace,
			wantRequestID:   goodReq,
			wantTraceSource: "",
		},
	}
	for _, tt := range tests {
		t.Run(tt.description, func(t *testing.T) {
			h := http.Header{}
			for k, v := range tt.headers {
				h.Set(k, v)
			}
			got := FromHeaderOrNew(h)

			assertID := func(name, want, actual string) {
				if want == "" {
					if actual == tt.headers[name] && tt.headers[name] != "" {
						t.Errorf("%s = %q, expected a regenerated id", name, actual)
					}
					if !validID(actual) {
						t.Errorf("%s = %q is not a valid generated id", name, actual)
					}
					return
				}
				if actual != want {
					t.Errorf("%s = %q, want %q", name, actual, want)
				}
			}
			assertID(headerTraceID, tt.wantTraceID, got.TraceID)
			assertID(headerRequestID, tt.wantRequestID, got.RequestID)
			if got.TraceSource != tt.wantTraceSource {
				t.Errorf("TraceSource = %q, want %q", got.TraceSource, tt.wantTraceSource)
			}
			if got.RequestSource != tt.wantRequestSource {
				t.Errorf("RequestSource = %q, want %q", got.RequestSource, tt.wantRequestSource)
			}
		})
	}
}

func TestFromHeaderOrNewTraceStart(t *testing.T) {
	now := time.Now().UTC()
	tests := []struct {
		description string
		header      string
		expected    func(got time.Time) bool
	}{
		{
			description: "absent trace-start defaults to ~now",
			header:      "",
			expected: func(got time.Time) bool {
				return !got.After(now.Add(time.Second)) && !got.Before(now.Add(-time.Second))
			},
		},
		{
			description: "valid past trace-start is preserved",
			header:      "2020-01-01T00:00:00Z",
			expected: func(got time.Time) bool {
				want, _ := time.Parse(time.RFC3339, "2020-01-01T00:00:00Z")
				return got.Equal(want)
			},
		},
		{
			description: "invalid trace-start falls back to ~now",
			header:      "not-a-timestamp",
			expected:    func(got time.Time) bool { return !got.Before(now.Add(-time.Second)) },
		},
		{
			description: "future trace-start is clamped to ~now",
			header:      now.Add(time.Hour).Format(time.RFC3339),
			expected:    func(got time.Time) bool { return !got.After(now.Add(time.Second)) },
		},
	}
	for _, tt := range tests {
		t.Run(tt.description, func(t *testing.T) {
			h := http.Header{}
			if tt.header != "" {
				h.Set(headerTraceStart, tt.header)
			}
			got := FromHeaderOrNew(h)
			if !tt.expected(got.TraceStart) {
				t.Errorf("TraceStart = %v not within expectation for %q", got.TraceStart, tt.header)
			}
		})
	}
}

func TestSaveToHeaderRoundTrip(t *testing.T) {
	orig := Trace{
		TraceID:       sampleTraceID,
		RequestID:     sampleRequestID,
		TraceSource:   sampleSource,
		RequestSource: sampleSource,
		TraceStart:    time.Now().UTC().Truncate(time.Second),
	}
	h := http.Header{}
	SaveToHeader(h, orig)

	got := FromHeaderOrNew(h)
	if got.TraceID != orig.TraceID || got.RequestID != orig.RequestID {
		t.Errorf("round-trip ids: got trace=%q req=%q, want trace=%q req=%q", got.TraceID, got.RequestID, orig.TraceID, orig.RequestID)
	}
	if got.TraceSource != orig.TraceSource {
		t.Errorf("round-trip TraceSource = %q, want %q", got.TraceSource, orig.TraceSource)
	}
	if !got.TraceStart.Equal(orig.TraceStart) {
		t.Errorf("round-trip TraceStart = %v, want %v", got.TraceStart, orig.TraceStart)
	}
}

func TestSaveToHeaderEmitsNoUnsafeBytes(t *testing.T) {
	// A Trace built from a poisoned inbound header must never re-emit CR/LF (or
	// other control bytes) through SaveToHeader — otherwise the outbound request
	// fails or smuggles a header downstream.
	poisoned := http.Header{}
	poisoned.Set(headerTraceID, "abc\r\nX-Evil: 1")
	poisoned.Set(headerRequestID, "req\r\nX-Evil: 2")
	poisoned.Set(headerTraceSource, "svc\r\nX-Evil: 3")
	trc := FromHeaderOrNew(poisoned)

	out := http.Header{}
	SaveToHeader(out, trc)
	for key, values := range out {
		for _, v := range values {
			if strings.ContainsAny(v, "\r\n\x00") {
				t.Errorf("header %q carries unsafe bytes after SaveToHeader: %q", key, v)
			}
		}
	}
}

type capturingRT struct{ req *http.Request }

func (c *capturingRT) RoundTrip(r *http.Request) (*http.Response, error) {
	c.req = r
	return &http.Response{StatusCode: http.StatusOK, Body: http.NoBody, Header: make(http.Header)}, nil
}

func TestClientMiddleware(t *testing.T) {
	t.Run("reuses trace id and mints a new request id when ctx has a trace", func(t *testing.T) {
		parent := Trace{TraceID: "trace-abc", RequestID: "req-parent", TraceSource: "svc"}
		cap := &capturingRT{}
		rt := ClientMiddleware(cap)

		req, _ := http.NewRequest(http.MethodGet, "http://example.invalid/", nil)
		req = req.WithContext(CtxWith(req.Context(), parent))
		resp, err := rt.RoundTrip(req)
		if err != nil {
			t.Fatalf("RoundTrip: %v", err)
		}
		resp.Body.Close()
		if got := cap.req.Header.Get(headerTraceID); got != "trace-abc" {
			t.Errorf("X-Trace-ID = %q, want the parent trace id", got)
		}
		if got := cap.req.Header.Get(headerRequestID); got == "" || got == "req-parent" {
			t.Errorf("X-Request-ID = %q, want a fresh sub-request id", got)
		}
	})

	t.Run("creates a fresh trace when ctx has none", func(t *testing.T) {
		cap := &capturingRT{}
		rt := ClientMiddleware(cap)
		req, _ := http.NewRequest(http.MethodGet, "http://example.invalid/", nil)
		resp, err := rt.RoundTrip(req)
		if err != nil {
			t.Fatalf("RoundTrip: %v", err)
		}
		resp.Body.Close()
		if got := cap.req.Header.Get(headerTraceID); !validID(got) {
			t.Errorf("X-Trace-ID = %q, want a generated valid id", got)
		}
	})
}

func TestServerMiddlewarePutsTraceInContext(t *testing.T) {
	var seen Trace
	var ok bool
	h := ServerMiddleware(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		seen, ok = FromCtx(r.Context())
	}))
	req, _ := http.NewRequest(http.MethodGet, "http://example.invalid/", nil)
	req.Header.Set(headerTraceID, "trace-from-header")
	h.ServeHTTP(nil, req)
	if !ok {
		t.Fatal("expected a Trace in the request context")
	}
	if seen.TraceID != "trace-from-header" {
		t.Errorf("TraceID = %q, want %q", seen.TraceID, "trace-from-header")
	}
}

func TestNewGeneratesUniqueIDs(t *testing.T) {
	const n = 10_000
	traces := make(map[string]struct{}, n)
	reqs := make(map[string]struct{}, n)
	for i := 0; i < n; i++ {
		tr := New()
		if _, dup := traces[tr.TraceID]; dup {
			t.Fatalf("duplicate TraceID at i=%d: %q", i, tr.TraceID)
		}
		if _, dup := reqs[tr.RequestID]; dup {
			t.Fatalf("duplicate RequestID at i=%d: %q", i, tr.RequestID)
		}
		traces[tr.TraceID] = struct{}{}
		reqs[tr.RequestID] = struct{}{}
	}
}

func TestValidIDIsAllocationFree(t *testing.T) {
	input := sampleTraceID
	if allocs := testing.AllocsPerRun(1000, func() { _ = validID(input) }); allocs != 0 {
		t.Errorf("validID allocated %v times per run, want 0", allocs)
	}
}

func BenchmarkFromHeaderOrNew(b *testing.B) {
	valid := http.Header{}
	valid.Set(headerTraceID, sampleTraceID)
	valid.Set(headerRequestID, sampleRequestID)

	invalid := http.Header{}
	invalid.Set(headerTraceID, "abc\r\nX-Evil: 1")

	absent := http.Header{}

	cases := map[string]http.Header{"present": valid, "absent": absent, "invalid": invalid}
	for name, h := range cases {
		b.Run(name, func(b *testing.B) {
			b.ReportAllocs()
			for i := 0; i < b.N; i++ {
				_ = FromHeaderOrNew(h)
			}
		})
	}
}

func BenchmarkValidID(b *testing.B) {
	input := sampleTraceID
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_ = validID(input)
	}
}

func BenchmarkNew(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_ = New()
	}
}

func BenchmarkNewUUID(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_ = newuuid()
	}
}

func BenchmarkSaveToHeader(b *testing.B) {
	trc := New()
	h := http.Header{}
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		SaveToHeader(h, trc)
	}
}

func BenchmarkClientMiddlewareRoundTrip(b *testing.B) {
	rt := ClientMiddleware(&capturingRT{})
	req, _ := http.NewRequest(http.MethodGet, "http://example.invalid/", nil)
	req = req.WithContext(CtxWith(context.Background(), New()))
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		resp, _ := rt.RoundTrip(req)
		resp.Body.Close()
	}
}

File: py/trace_test.py
import unittest
from datetime import datetime, timedelta, timezone
from trace import Trace, _resolve_id, _resolve_source, _valid_id


class TestValidId(unittest.TestCase):
    def test_cases(self):
        cases = [
            {"description": "canonical uuid", "input": "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31", "expected": True},
            {"description": "service-prefixed id", "input": "trace_018f3a1c2b4d", "expected": True},
            {"description": "dotted id", "input": "svc.abc.123", "expected": True},
            {"description": "max length accepted", "input": "a" * 200, "expected": True},
            {"description": "empty rejected", "input": "", "expected": False},
            {"description": "one over max rejected", "input": "a" * 201, "expected": False},
            {"description": "CRLF rejected", "input": "abc\r\nX-Evil: 1", "expected": False},
            {"description": "newline rejected", "input": "abc\ndef", "expected": False},
            {"description": "space rejected", "input": "abc def", "expected": False},
            {"description": "control byte rejected", "input": "abc\x00def", "expected": False},
            {"description": "non-ascii rejected", "input": "abcé", "expected": False},
            {"description": "slash rejected", "input": "a/b", "expected": False},
        ]
        for c in cases:
            with self.subTest(c["description"]):
                self.assertEqual(_valid_id(c["input"]), c["expected"])


class TestResolveId(unittest.TestCase):
    def test_cases(self):
        cases = [
            {"description": "valid id propagated", "input": "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31", "propagate": True},
            {"description": "absent id generated", "input": "", "propagate": False},
            {"description": "poisoned id regenerated", "input": "abc\r\nX-Evil: 1", "propagate": False},
            {"description": "over-long id regenerated", "input": "a" * 201, "propagate": False},
        ]
        for c in cases:
            with self.subTest(c["description"]):
                got = _resolve_id(c["input"])
                if c["propagate"]:
                    self.assertEqual(got, c["input"])
                else:
                    self.assertNotEqual(got, c["input"])
                    self.assertTrue(_valid_id(got), f"generated id {got!r} is not valid")


class TestResolveSource(unittest.TestCase):
    def test_cases(self):
        cases = [
            {"description": "empty source stays empty", "input": "", "expected": ""},
            {"description": "valid source kept", "input": "runpod-graphql", "expected": "runpod-graphql"},
            {"description": "source at max length kept", "input": "a" * 64, "expected": "a" * 64},
            {"description": "poisoned source dropped to empty", "input": "svc\r\nX-Evil: 1", "expected": ""},
            {"description": "spaced source dropped to empty", "input": "not a slug", "expected": ""},
            {"description": "over-length source dropped to empty", "input": "a" * 65, "expected": ""},
            {"description": "charset-valid but over-length source dropped", "input": "a" * 200, "expected": ""},
        ]
        for c in cases:
            with self.subTest(c["description"]):
                self.assertEqual(_resolve_source(c["input"]), c["expected"])


class TestFromHeaders(unittest.TestCase):
    def test_cases(self):
        good_trace = "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31"
        good_req = "018f3a1c-2b4d-7e8f-9a0b-1c2d3e4f5061"
        cases = [
            {
                "description": "valid ids propagated",
                "headers": {"X-Trace-ID": good_trace, "X-Request-ID": good_req, "X-Trace-Source": "main-ui"},
                "want_trace": good_trace,
                "want_request": good_req,
                "want_trace_source": "main-ui",
            },
            {
                "description": "absent ids generated",
                "headers": {},
                "want_trace": None,
                "want_request": None,
                "want_trace_source": "",
            },
            {
                "description": "poisoned trace id regenerated, valid request id kept",
                "headers": {"X-Trace-ID": "abc\r\nX-Evil: 1", "X-Request-ID": good_req},
                "want_trace": None,
                "want_request": good_req,
                "want_trace_source": "",
            },
            {
                "description": "poisoned source dropped to empty",
                "headers": {"X-Trace-ID": good_trace, "X-Request-ID": good_req, "X-Trace-Source": "svc\r\nX-Evil"},
                "want_trace": good_trace,
                "want_request": good_req,
                "want_trace_source": "",
            },
        ]
        for c in cases:
            with self.subTest(c["description"]):
                t = Trace.from_headers(dict(c["headers"]))
                if c["want_trace"] is None:
                    self.assertTrue(_valid_id(t.trace_id))
                    self.assertNotIn("\r", t.trace_id)
                else:
                    self.assertEqual(t.trace_id, c["want_trace"])
                if c["want_request"] is None:
                    self.assertTrue(_valid_id(t.request_id))
                else:
                    self.assertEqual(t.request_id, c["want_request"])
                self.assertEqual(t.trace_source, c["want_trace_source"])


class TestSaveToHeaders(unittest.TestCase):
    def test_writes_go_five_headers_and_fresh_request_id(self):
        t = Trace.from_headers({"X-Trace-ID": "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31"})
        original_request_id = t.request_id
        headers: dict[str, str] = {}
        t.save_to_headers(headers)

        for key in ("X-Trace-ID", "X-Request-ID", "X-Trace-Start", "X-Trace-Source", "X-Request-Source"):
            self.assertIn(key, headers)

        self.assertEqual(headers["X-Trace-ID"], t.trace_id)
        # Regression: X-Request-ID must be a plain string, not a tuple (trailing-comma bug).
        self.assertIsInstance(headers["X-Request-ID"], str)
        # A fresh sub-request id, distinct from the inbound one.
        self.assertNotEqual(headers["X-Request-ID"], original_request_id)

    def test_emits_no_unsafe_bytes_from_poisoned_trace(self):
        # Every inbound header poisoned, including X-Trace-Start — the timestamp is
        # the header that previously leaked CR/LF straight through save_to_headers.
        cases = [
            {"description": "poisoned trace id", "header": "X-Trace-ID", "value": "abc\r\nX-Evil: 1"},
            {"description": "poisoned trace source", "header": "X-Trace-Source", "value": "svc\r\nX-Evil: 2"},
            {"description": "poisoned request id", "header": "X-Request-ID", "value": "req\r\nX-Evil: 3"},
            {"description": "poisoned trace start", "header": "X-Trace-Start", "value": "2020-01-01T00:00:00Z\r\nX-Evil: 4"},
        ]
        for c in cases:
            with self.subTest(c["description"]):
                t = Trace.from_headers({c["header"]: c["value"]})
                headers: dict[str, str] = {}
                t.save_to_headers(headers)
                for key, value in headers.items():
                    self.assertNotIn("\r", value, f"{key} carries CR")
                    self.assertNotIn("\n", value, f"{key} carries LF")

    def test_malformed_trace_start_falls_back_to_now(self):
        cases = [
            {"description": "non-RFC3339 falls back", "value": "not-a-date"},
            {"description": "CRLF-poisoned falls back", "value": "2020-01-01T00:00:00Z\r\nX-Evil: 1"},
        ]
        for c in cases:
            with self.subTest(c["description"]):
                t = Trace.from_headers({"X-Trace-ID": "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31", "X-Trace-Start": c["value"]})
                # The stored value re-parses as an RFC3339 timestamp (the fallback
                # to now), never the poisoned input.
                self.assertNotIn("\r", t.trace_start)
                parsed = datetime.fromisoformat(t.trace_start.replace("Z", "+00:00"))
                delta = abs((datetime.now(timezone.utc) - parsed).total_seconds())
                self.assertLess(delta, 5, f"trace_start {t.trace_start!r} is not ~now")

    def test_valid_trace_start_is_preserved(self):
        t = Trace.from_headers(
            {"X-Trace-ID": "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31", "X-Trace-Start": "2020-01-01T00:00:00Z"}
        )
        parsed = datetime.fromisoformat(t.trace_start.replace("Z", "+00:00"))
        self.assertEqual((parsed.year, parsed.month, parsed.day), (2020, 1, 1))

    def test_future_trace_start_is_clamped_to_now(self):
        # A caller-supplied future timestamp is clamped to now (matching Go), so it
        # never drives the trace's start time.
        future = (datetime.now(timezone.utc) + timedelta(hours=1)).strftime("%Y-%m-%dT%H:%M:%SZ")
        t = Trace.from_headers(
            {"X-Trace-ID": "5f9c2e6a-1b3d-4c8e-9a0f-2b7c6d5e4f31", "X-Trace-Start": future}
        )
        parsed = datetime.fromisoformat(t.trace_start.replace("Z", "+00:00"))
        delta = abs((datetime.now(timezone.utc) - parsed).total_seconds())
        self.assertLess(delta, 5, f"future trace_start {t.trace_start!r} was not clamped to now")


class TestNewRegression(unittest.TestCase):
    def test_new_does_not_raise(self):
        # Regression: new() used datetime.datetime.now() and raised AttributeError.
        t = Trace.new()
        self.assertTrue(_valid_id(t.trace_id))
        self.assertTrue(_valid_id(t.request_id))

    def test_new_generates_unique_ids(self):
        seen = set()
        for _ in range(1000):
            seen.add(Trace.new().trace_id)
        self.assertEqual(len(seen), 1000)


if __name__ == "__main__":
    unittest.main()
